### DISCUSSION TITLE
Secure Communication Between Actions and App

### DISCUSSION THREAD

**Discussion by romeara** [community/community#24700] [Apps, API and Webhooks]
<p>I maintain a GitHub App with the following flow:</p>
<ul>
<li>A GitHub Action is added to a workflow</li>
<li>This action runs a task, and uploads results to the App</li>
<li>The App performs more complex analysis and interacts with GitHub</li>
</ul>
<p>Ideally, I’d like a way to verify that the upload in the second step comes from a GitHub process - basically, a shared secret, similar to how I can verify webhook payloads. I don’t want to require users of the App to manually configure an organization-level secret, and most are hesitant to grant the “secrets” permission required to setup such a secret via API upon installation. It would be great if Apps automatically had a shared secret associated with an installation, or similar.</p>
<p>Does such functionality already exist, or if not, is there somewhere other than here I should be starting a request/discussion about servicing this use case?</p>

**Comment by shrink** [comment, marked as answer]
<p>I don’t think this functionality exists, but that’s an interesting idea… and I think there may be options to achieve this now.</p>
<p>The first option is to use the <code>GITHUB_TOKEN</code> associated with the Run, as this token will remain active while the Run is in progress. The upload endpoint of your App would accept the <code>GITHUB_TOKEN</code> from the Action and use it to verify access to the target repository.</p>
<pre><code class="lang-auto">POST https://example.com/uploads

{
  "repository": "examples/example-repository",
  "token": "${{ github.token }}",
  "data": {}, // your upload data
}
</code></pre>
<p>Then you can use the 200 or 401 response of this request to pass or fail the check.</p>
<pre><code class="lang-bash">curl --request GET \
  --url https://api.github.com/repos/:repository \
  --header 'Authorization: token :github_token' \
  --header 'Content-Type: application/json'
</code></pre>
<p>The downside of this approach is that your validation of an upload must be synchronous because the token will expire the moment the Run completes – and so doing it in the background is not an option.</p>
<p>Another option could be to implement an intermediary step: accept uploads without any authentication, <em>but</em> hold the uploads for processing until a GitHub Webhook provides you information required to determine the upload was genuine.</p>
<p>For example, <a href="https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#workflow_run" rel="noopener nofollow ugc"><code>workflow_run</code></a> could be used in the following way:</p>
<ol>
<li>Your Action, in a Workflow, uploads a result to your App with a commit, repository or Check Run identifier – whichever is most useful to you</li>
<li>Your App adds the uploaded result to a “queue” of uploads “pending validation”</li>
<li>Your App listens for a <code>workflow_run</code> Webhook event</li>
<li>When a <code>workflow_run</code> Webhook event arrives, the App extracts the commit information from the signed Webhook and uses it to approve any matching uploads</li>
</ol>
<p>For example, if I submit an upload for <code>examples/my-repository@a5b457d</code> and then you receive a <code>workflow_run</code> webhook for that commit, you can say with reasonable confidence that the upload came from GitHub. You could take it a step further and use the Webhook information to retrieve the Workflow yaml and verify that it did indeed use your Action to upload the file – probably overkill, but helpful if you’d like to also protect against authorised users misbehaving.</p>
<p>Does that help, or do you have any considerations these options miss?</p>

**Comment by romeara** [comment]
<p>The <code>workflow_run</code> approach is certainly interesting, and may actually help a lot, thank you! (though I’d still advocate for a little more formalized approach like I outlined)</p>
<p>Using <code>GITHUB_TOKEN</code> had occurred to me, but I was hesitant to send a secret like that in a request body or header, outside the system it’s meant to be encapsulated by. If there’s an argument for this being ok, we actually already have places in accepting the upload that attempts verification for systems that allow for it that such an operation could be slotted into</p>

**Comment by shrink** [comment]
<aside class="quote no-group" data-username="romeara" data-post="3" data-topic="201330">
<div class="title">
<div class="quote-controls"></div>
 romeara:</div>
<blockquote>
<p>(though I’d still advocate for a little more formalized approach like I outlined)</p>
</blockquote>
</aside>
<p>I agree! Hopefully, someone from GitHub sees this but their <a href="https://github.com/github/roadmap/projects/1?card_filter_query=label%3Aactions" rel="noopener nofollow ugc">Roadmap</a> is very full so it’s probably worth moving forward with a workaround if you need to deliver this in the next few quarters.</p>
<aside class="quote no-group quote-modified" data-username="romeara" data-post="3" data-topic="201330">
<div class="title">
<div class="quote-controls"></div>
 romeara:</div>
<blockquote>
<p>…I was hesitant to send a secret like that in a request body or header, outside the system it’s meant to be encapsulated by. If there’s an argument for this being ok…</p>
</blockquote>
</aside>
<p>I think there are arguments in favour and arguments against. The key argument against this is that these tokens are not designed to be used externally and a security audit from an especially cautious customer might flag it as a concern, “why is this Action exfiltrating our tokens?”.</p>
<p>Personally, though, I think there is a strong justification for using these tokens externally <em>if</em> done so with security in mind. The tokens are described as a way to “<a href="https://docs.github.com/en/actions/reference/authentication-in-a-workflow" rel="noopener nofollow ugc">authenticate on behalf of GitHub Actions</a>” which <em>is</em> the intent of your implementation. The permissions configuration for these tokens enables a Workflow administrator (i.e: your customer) to confidently restrict the scope of the token down to the <a href="https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token" rel="noopener nofollow ugc"><em>same level that a GitHub App has by default</em></a>.</p>
<pre><code class="lang-yaml">jobs:
  do-app-task:
    runs-on: ubuntu-latest
    permissions:
      metadata: read
    steps:
      - uses: app/action@v1
        with:
          token: ${{ github.token }}
</code></pre>
<p>From the <a href="https://docs.github.com/en/rest/reference/permissions-required-for-github-apps" rel="noopener nofollow ugc">GitHub Apps Permissions documentation</a> (emphasis mine):</p>
<blockquote>
<p>The metadata permission provides access to a collection of read-only endpoints with metadata for various resources. <strong>These endpoints do not leak sensitive private repository information.</strong></p>
</blockquote>
<p>My conclusion is that it’s justifiable, the key points that lead me to that conclusion:</p>
<ol>
<li>The tokens immediately expire at the end of a Workflow Run, this is behaviour controlled by GitHub and so a customer is not dependent on you to be a responsible custodian of the tokens</li>
<li>GitHub are very conscious of the threat that these tokens present, so much so there is <a href="https://securitylab.github.com/research/" rel="noopener nofollow ugc">security research published about secure use of these tokens</a> and this approach does not contradict any of that advice</li>
<li>The token permissions can be restricted <em>by the Workflow administrator</em> to provide the same level of permissions that a GitHub App has already</li>
<li>The <code>metadata</code> endpoints available are explicitly described by GitHub as not leaking sensitive data – something that should give users confidence</li>
</ol>

**Comment by romeara** [comment]
<p>That is a very comprehensive analysis, thank you! based on what you’d provided, I’m leaning towards using the token, at least as a first iteration to improve upon the current situation.</p>

**Comment by wparad** [comment]
<p>You definitely don’t want to use the <code>GITHUB_TOKEN</code> like this it isn’t secure. There are lots of reasons, but the base of it is, anyone in all of github can create a token that may have the same attributes as the <code>GITHUB_TOKEN</code> and therefor indistiguishable.</p>
<p>It isn’t official yet, but it may be the case you can generate a JWT using this action</p>
<pre><code class="lang-auto">curl -H "Authorization: bearer $ACTIONS_ID_TOKEN_REQUEST_TOKEN" "$ACTIONS_ID_TOKEN_REQUEST_URL" | jq -r '.value' &gt; $AWS_WEB_IDENTITY_TOKEN_FILE
</code></pre>
<p>And from there pass the JWT to your app to verify it. If the token doesn’t have what you need you can configure it further using third party services such as Vault or Authress.</p>
